### PR TITLE
Add method to serialize collection/catalog

### PR DESCRIPTION
**Pull request by andersy005**
Towards #166

### CODE REVIEW

File: intake_esm/core.py [hunk 1]
```diff
@@ -106,13 +106,49 @@ def search(self, **query):
...
+            directory = Path(directory)
+            directory.mkdir(parents=True, exist_ok=True)
+            csv_file_name = directory / csv_file_name
+            json_file_name = directory / json_file_name
+
+        print(f'Writing csv catalog to: {csv_file_name}')
+        self.df.to_csv(csv_file_name, compression='gzip', index=False)
+        print(f'Writing ESM collection json file to: {json_file_name}')
+        with open(json_file_name, 'w') as outfile:
+            json.dump(self._col_data, outfile)
```
Comment (jhamman) [inline intake_esm/core.py]: Don't you need to set the `catalog_file` field in the spec somewhere in here? Am I missing some magic?

File: tests/cmip6/test_cmip6.py [hunk 2]
```diff
@@ -119,3 +120,18 @@ def test_load_esmcol_remote():
...
+            'https://raw.githubusercontent.com/NCAR/intake-esm-datastore/master/catalogs/pangeo-cmip6.json'
+        )
+        col_subset = col.search(
+            source_id='BCC-ESM1', grid_label='gn', table_id='Amon', experiment_id='historical'
+        )
+
+        col_subset.serialize(name='cmip6_bcc_esm1', directory=local_store)
+
+        col = intake.open_esm_datastore(f'{local_store}/cmip6_bcc_esm1.json')
+        assert len(col.df) > 0
```
Comment (jhamman) [inline tests/cmip6/test_cmip6.py]: I think you want this to be more of an explicit check. `assert col.df.equals(col_subset.df)` should do it, yes?

File: intake_esm/core.py (hunk 1 above)
Comment (andersy005) [reply intake_esm/core.py]: Good catch!! I totally missed this :)

File: tests/cmip6/test_cmip6.py (hunk 2 above)
Comment (andersy005) [reply tests/cmip6/test_cmip6.py]: 👍

**Comment by andersy005** [comment]
> You could do that by including the dataframe in the json file.

This is very interesting! Considering the size of the some of the csv files (>=50MB compressed), wouldn't this make the single json file cumbersome?

File: tests/cmip6/test_cmip6.py [hunk 3]
```diff
@@ -119,3 +121,21 @@ def test_load_esmcol_remote():
...
+        )
+        col_subset = col.search(
+            source_id='BCC-ESM1', grid_label='gn', table_id='Amon', experiment_id='historical'
+        )
+
+        name = 'cmip6_bcc_esm1'
+        col_subset.serialize(name=name, directory=local_store)
+
+        col = intake.open_esm_datastore(f'{local_store}/cmip6_bcc_esm1.json')
+        pd.testing.assert_frame_equal(col_subset.df.reset_index(drop=True), col.df)
```
Comment (jhamman) [inline tests/cmip6/test_cmip6.py]: the `reset_index` piece is worrying here. Why is this needed?

**Comment by jhamman** [comment]
> > You could do that by including the dataframe in the json file.
> 
> This is very interesting! Considering the size of the some of the csv files (>=50MB compressed), wouldn't this make the single json file cumbersome?

Let's move this to a new issue in the spec repo (I'll open this). It isn't something we should consider for this PR.

File: tests/cmip6/test_cmip6.py (hunk 3 above)
Comment (andersy005) [reply tests/cmip6/test_cmip6.py]: > the reset_index piece is worrying here. Why is this needed?

`col.search()` returns dataframe with original indices. When saving the subset, I am not keeping these original indices. As a result, `pd.testing.assert_frame_equal()` fails when used without `reset_index()`:


```python
In [3]: col = intake.open_esm_datastore("https://raw.githubusercontent.com/NCAR/intake-esm-datastore/master/catalogs/pangeo-cmi
   ...: p6.json")                                                                                                              

In [4]: col.df.head()                                                                                                          
Out[4]: 
  activity_id institution_id source_id  ... grid_label                                             zstore dcpp_init_year
0  AerChemMIP            BCC  BCC-ESM1  ...         gn  gs://cmip6/AerChemMIP/BCC/BCC-ESM1/piClim-CH4/...            NaN
1  AerChemMIP            BCC  BCC-ESM1  ...         gn  gs://cmip6/AerChemMIP/BCC/BCC-ESM1/piClim-CH4/...            NaN
2  AerChemMIP            BCC  BCC-ESM1  ...         gn  gs://cmip6/AerChemMIP/BCC/BCC-ESM1/piClim-CH4/...            NaN
3  AerChemMIP            BCC  BCC-ESM1  ...         gn  gs://cmip6/AerChemMIP/BCC/BCC-ESM1/piClim-CH4/...            NaN
4  AerChemMIP            BCC  BCC-ESM1  ...         gn  gs://cmip6/AerChemMIP/BCC/BCC-ESM1/piClim-CH4/...            NaN

[5 rows x 10 columns]

In [5]: col_subset = col.search(source_id="BCC-ESM1", grid_label="gn", table_id="Amon", experiment_id="historical")            

In [6]: col_subset.df.head()                                                                                                   
Out[6]: 
     activity_id institution_id source_id  ... grid_label                                             zstore dcpp_init_year
5416        CMIP            BCC  BCC-ESM1  ...         gn  gs://cmip6/CMIP/BCC/BCC-ESM1/historical/r1i1p1...            NaN
5417        CMIP            BCC  BCC-ESM1  ...         gn  gs://cmip6/CMIP/BCC/BCC-ESM1/historical/r1i1p1...            NaN
5418        CMIP            BCC  BCC-ESM1  ...         gn  gs://cmip6/CMIP/BCC/BCC-ESM1/historical/r1i1p1...            NaN
5419        CMIP            BCC  BCC-ESM1  ...         gn  gs://cmip6/CMIP/BCC/BCC-ESM1/historical/r1i1p1...            NaN
5420        CMIP            BCC  BCC-ESM1  ...         gn  gs://cmip6/CMIP/BCC/BCC-ESM1/historical/r1i1p1...            NaN
```

Comment (andersy005) [reply tests/cmip6/test_cmip6.py]: Should we keep these original indices?

Comment (jhamman) [reply tests/cmip6/test_cmip6.py]: I think we should either save the original indicies OR have `search()` return reset its own indicies. Ultimately, I think we just want to be internally consistent here.

Comment (andersy005) [reply tests/cmip6/test_cmip6.py]: I concur! I added the `reset_index()` step as part of `search()` for internal consistency

**Comment by andersy005** [comment]
@jhamman, this is ready for another look. If everything looks good to you, please go ahead and merge it.

**Comment by jhamman** [comment]
Nice work @andersy005!

**Comment by andersy005** [comment]
Thank you, @jhamman!
